# Web Scraping

## Part \#3: XML and Semi-Structured Data

### Where do data scientists get their data?

A data scientist needs sources for data to do his or her work. While you don't need this level of detail on the AP test, here are is some information about the types of data available to data scientists:

* **Unstructured data:** Most data available on the web is unstructured data. Image files, sound files, video files, text files, and HTML files are all examples of unstructured data. These are some of the richest sources of data available, but they are also difficult to process (search, sort, classify, analyze, summarize etc). For example, while you were easily able to write algorithms in a previous chapter to apply image filters to the pixels of an image, it is very difficult to write functions to classify the objects featured in image files (people, trees, dogs, cats, mountains, etc). Similarly, while a web browser can effectively display an HTML file, it is not easy to write an algorithm to give a short summary of the content of the webpage. You can certainly extract useful data from an HTML file, but you just saw that it takes a great deal of effort using a module such as **Beautiful Soup**. In fact, the name of the module describes the World Wide Web: it is a beautiful soup of unstructured data. Data scientists often spend a great deal of time finding their data. To read more, here's a great blog post on the topic: https://www.dataquest.io/blog/web-scraping-beautifulsoup/


* **Semi-structured data:** You already have experience with semi-structured data: a *.csv* file is an example of semi-structured data. This is data that supports automated processing of its contents, such as we saw with Pandas during the our chapter on Open Data. In other courses, you may learn about [JSON](https://www.w3schools.com/js/js_json_intro.asp), and in this notebook you will learn about [XML](https://www.w3schools.com/xml/xml_whatis.asp). As you will soon see, the beauty of XML is that you can work with data in an automated way.


* **Structured data:** This is data that is stored in a *database*. Organizations such as corporations, governments, and universities will have servers dedicated to their databases and database software. The data stored in a database is similar to what you have seen in *.csv* files, but has some additional structure. We will not work with databases (structured data) in this class, but they are a great source of information. But if you want to learn more, you may want look up the term *relational database* or *SQL*.  

### What is XML?

XML stands for eXtensible Markup Language. While HTML is meant to display webpages, XML is meant to store/transport/describe data. Humans can read and understand XML, and computers can also process XML in an automated way.

**Task \#1:** Read the following pages from W3Schools, then answer the questions below:

  * XML Introduction: https://www.w3schools.com/xml/xml_whatis.asp

  * XML Tutorial: https://www.w3schools.com/xml/default.asp

**Question \#1:** What does the "extensible" in "extensible markup language" mean?

Your Answer: Extensible means that the structure of the tags will always be the same but the data between the tags can be changed at any other time without changing anything else.

**Question \#2:** Suppose you are trying to write an algorithm to process data in an automated way. Why would you prefer for your algorithm to work with *extensible* data? 

Your Answer: Working with extensible data means that the user doesn't have to put much thought into how the data is structured because the data already has a semi-permanent structure. An algorithm only has to change specific parts of extensible data and doesn't have to make any decisions on whether to modify anything else.

### Chicago Weather Data

Before you read any further, visit the page http://w1.weather.gov/xml/current_obs/KORD.xml and look at the data on the page. This is a feed of current weather conditions at Chicago's O'Hare airport. You can use this XML document to create a webpage or app that always knows the most up-to-date weather in Chicago. 

Let's take a look at the XML source for this feed:

In [1]:
from bs4 import BeautifulSoup  
from urllib.request import urlopen

xml_page = urlopen("http://w1.weather.gov/xml/current_obs/KORD.xml")   # Opens whatever page we are requesting
bs_obj = BeautifulSoup(xml_page, 'xml')

print(bs_obj.prettify())

<?xml version="1.0" encoding="utf-8"?>
<?xml-stylesheet href="latest_ob.xsl" type="text/xsl"?>
<current_observation version="1.0" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="http://www.weather.gov/view/current_observation.xsd">
 <credit>
  NOAA's National Weather Service
 </credit>
 <credit_URL>
  http://weather.gov/
 </credit_URL>
 <image>
  <url>
   http://weather.gov/images/xml_logo.gif
  </url>
  <title>
   NOAA's National Weather Service
  </title>
  <link>
   http://weather.gov
  </link>
 </image>
 <suggested_pickup>
  15 minutes after the hour
 </suggested_pickup>
 <suggested_pickup_period>
  60
 </suggested_pickup_period>
 <location>
  Chicago, Chicago-O'Hare International Airport, IL
 </location>
 <station_id>
  KORD
 </station_id>
 <latitude>
  41.97972
 </latitude>
 <longitude>
  -87.90444
 </longitude>
 <observation_time>
  Last Updated on Apr 17 2019, 5:51 pm CDT
 </observation_time>
 <ob

If you want to extract the current temperature from this data, just run the code below:

In [2]:
current_temp = bs_obj.find('temp_f').getText()
print(current_temp)

74.0


**Question \#3:** What will you probably notice about the output from the cell directly above if you were to re-run the code in this notebook at a later date?

Your Answer: The answer may be different because temperature changes all the time


**Question \#4:** Where was this temperature measurement taken? To answer this question, you must write code that uses the latitude and longitude data in the XML above to create a tuple.  Give your answer as a tuple in the form (latitude,longitude)

Location Tuple: ('41.97972', '-87.90444')


In [3]:
# Your code here
lat = bs_obj.find('latitude').getText() # find latitude
lon = bs_obj.find('longitude').getText() # find longitude
lat = float(lat) # turn into number
lon = float(lon) # turn into number
coords = (lat,lon) # turn this into a tuple
print(coords)

(41.97972, -87.90444)


**Task \#2** Use the Google Maps API to create a marker map with the location of the coordinates you found in the previous question.

In [5]:
# Import the gmaps python module and load in your API Key:
import gmaps

gmaps.configure(api_key="AIzaSyCLla6Q7krE9xNg6SnNMoGNIzjCLddE9EU") # you need to put your API key here in place of the X's

from ipywidgets.embed import embed_minimal_html # Allows us to create a separte file for the Google Maps

locations = []

locations.append(coords)

markers = gmaps.marker_layer(locations)

markermap = gmaps.Map()

markermap = gmaps.Map()
markermap.add_layer(markers)

embed_minimal_html('MarkerMap1.html', views=[markermap])
print("*** Check your 'webscraping part 3' folder to find the new HTML file. ***")

ModuleNotFoundError: No module named 'gmaps'

**Question \#5:** What does the following function, *tag_extractor(url, tag)*, do? Some structure is provided below to help you answer this question:

Your Answer:  "tag_extractor" takes a link and a tag as input, visits the link, and returns the data within the tag.

* _What is the purpose of the function?_
    * Your answer: to quickly find the data under a specific webpages's specific tag
* _What purpose does the parameter **url** serve in the function?_
    * Your answer: "url" allows the user to type in a web link to serve as the web page where the data is being extracted from
* _What purpose does the parameter **tag** serve in the function?_
    * Your answer: "tag" allows the user to find a specific XML tag to extract the data from
* _What information is being returned by the function?_  
    * Your answer: The data under a specific tag from a specific webpage

In [9]:
def tag_extractor(url, tag):    
    from bs4 import BeautifulSoup  
    from urllib.request import urlopen

    xml_page = urlopen(url)   #opens whatever page we are requesting
    bs_obj = BeautifulSoup(xml_page, 'xml')
    
    return bs_obj.find(tag).getText()

tag_extractor('http://w1.weather.gov/xml/current_obs/KORD.xml', 'temp_f')

'74.0'

**Question \#6:** How can *tag_extractor(url, tag)* be considered an abstraction that helps to manage the complexity of a computer program? 

Your Answer: If I want to compare the weather from one internet source to another, I can run this function twice in order to find the temperature on each website. With this function, I need to run two lines of code to perform this task. Without this function, I need to run twelve lines of code. The function is considered an abstraction because it lessens the time and space I need to perform this task.

**Task \#3:** Use *tag_extractor(url, tag)* to determine the date/time of the most recent temperature measurement.

In [10]:
# Your code here
tag_extractor('http://w1.weather.gov/xml/current_obs/KORD.xml', 'observation_time')

'Last Updated on Apr 17 2019, 5:51 pm CDT'

### HTML as Output

**Question \#7:** What is the purpose of the function *html_output()*? What kind of data does html_output() produce as output?

Your Answer: *html_output* finds the current temperature from O'Hare and then creates an HTML file that displays this temperature. *html_output()* produces an HTML file as output.


In [11]:
# define the function: 
def html_output():    
    output_string = """
    <html>
    <head>
    <meta http-equiv="refresh" content="3600">
        <style>
            body {
                background-color: #BBBBBB; 
                text-align: center;        
            }
        </style>
    </head>

    <body>
    <h1>Chicago Weather</h1>
    <p> The current temperature in Chicago is 
    """

    output_string += tag_extractor('http://w1.weather.gov/xml/current_obs/KORD.xml', 'temp_f')

    output_string += """
    degrees Fahrenheit.
    </p>
    <br>

    </body>
    </html>
    """

    html_file= open("output/O'Hare Temperature.html","w")
    html_file.write(output_string)
    html_file.close()
    
# now call the function: 

html_output()
print("*** Check your 'Web Scraping Part 3' folder to find the new HTML file. ***")
type(html_output())

*** Check your 'Web Scraping Part 3' folder to find the new HTML file. ***


NoneType

**Task \#4:** Create your own version of *html_output()* that includes the date/time of the most recent temperature measurement as well as two other measurements to your output *.html* document. 

In [12]:
# Your code here
# define the function: 
def html_file():    
    output1 = """
    <html>
    <head>
        <style>
            body {
                background-color: #BBBBBB; 
                text-align: center;        
            }
        </style>
    </head>

    <body>
    <h1>Chicago Data</h1>
    <p> The temperature in Chicago was
    """

    output1 += tag_extractor('http://w1.weather.gov/xml/current_obs/KORD.xml', 'observation_time')

    output1 += """
    
    .
    </p>
    <br>
    <p> The current windspeed in Chicago is 
    """
    
    output1 += tag_extractor('http://w1.weather.gov/xml/current_obs/KORD.xml', 'wind_mph')
    
    output1 += """
    
    mph.
    </p>
    <br>
    <p> The current visiblity is
    """
    
    output1 += tag_extractor('http://w1.weather.gov/xml/current_obs/KORD.xml', 'visibility_mi')
    
    output1 += """
    
    miles.
    </p>
    <br>
    </body>
    </html>
    """

    html_file= open("output/O'Hare Data.html","w")
    html_file.write(output1)
    html_file.close()
    
# now call the function: 
html_file()
print("*** Check your 'Web Scraping Part 3' folder to find the new HTML file. ***")

*** Check your 'Web Scraping Part 3' folder to find the new HTML file. ***


**Task \#5:** This is a multistep task: 

1) Read this [w3schools documentation](https://www.w3schools.com/tags/att_meta_http_equiv.asp)

2) Add the HTML `<meta http-equiv="refresh" content="3600">` between the head tags in your html_output() function

3) Read (but do not run) the code in the cell below: 

In [13]:
running = False
import time


while running:
    html_output()
    time.sleep(3600)

**Question \#9:** If you were to change *running* to *True* in the code cell above, then running this cell would start an infinite loop. What would the purpose be of an infinite loop and the HTML above, `<meta http-equiv="refresh" content="3600">`? What would it allow you to do? 

Your Answer: The purpose of this infinite loop would be to update the webpage without the user having to do so manually. The code above would run the function every hour, therefore updating the webpage every hour, similar to how a weather webpage refreshes every so often.


### Experiment and Explore

**Task \#6:** With any extra time this period, go to http://w1.weather.gov/xml/current_obs/KORD.xml, but change the **KORD** portion of this URL to another ICAO airport code for an airport in the United States or its territories. Experiment. Show the results of your experimentation in an HTML output file, and produce a Google Maps Marker Map in this notebook that includes the location you chose to explore. 

  * Airport codes in the lower-48 states: https://en.wikipedia.org/wiki/List_of_airports_by_ICAO_code:_K

  * All airport codes in the United States and territories: https://en.wikipedia.org/wiki/List_of_airports_in_the_United_States#Lists_by_ICAO_location_indicator

In [19]:
# Your code here
# define the function: 
def html_experiment():    
    output1 = """
    <html>
    <head>
        <style>
            body {
                background-color: #AABBCC; 
                text-align: center;        
            }
        </style>
    </head>

    <body>
    <h1>Austin Data</h1>
    <p> The temperature in Austin was
    """

    output1 += tag_extractor('http://w1.weather.gov/xml/current_obs/KAUS.xml', 'observation_time')

    output1 += """
    
    .
    </p>
    <br>
    <p> The current windspeed in Austin is 
    """
    
    output1 += tag_extractor('http://w1.weather.gov/xml/current_obs/KAUS.xml', 'wind_mph')
    
    output1 += """
    
    mph.
    </p>
    <br>
    <p> The current visiblity is
    """
    
    output1 += tag_extractor('http://w1.weather.gov/xml/current_obs/KAUS.xml', 'visibility_mi')
    
    output1 += """
    
    miles.
    </p>
    <br>
    </body>
    </html>
    """

    html_experiment= open("output/Austin Experiment.html","w")
    html_experiment.write(output1)
    html_experiment.close()
    
# now call the function: 
html_experiment()
print("*** Check your 'Web Scraping Part 3' folder to find the new HTML file. ***")

*** Check your 'Web Scraping Part 3' folder to find the new HTML file. ***


In [21]:
# Your code here
lat = tag_extractor('http://w1.weather.gov/xml/current_obs/KAUS.xml', 'latitude')
lon = tag_extractor('http://w1.weather.gov/xml/current_obs/KAUS.xml', 'longitude')
lat = float(lat) # turn into number
lon = float(lon) # turn into number
coords = (lat,lon) # turn this into a tuple
print(coords)

(30.18304, -97.67987)


In [22]:
# Import the gmaps python module and load in your API Key:
import gmaps
gmaps.configure(api_key="AIzaSyCLla6Q7krE9xNg6SnNMoGNIzjCLddE9EU")

from ipywidgets.embed import embed_minimal_html # Allows us to create a separte file for the Google Maps

locations = []

locations.append(coords)

markers = gmaps.marker_layer(locations)    # Create markers for each tuple/coordinate
markermap = gmaps.Map()                         # Create a GMap variable
markermap.add_layer(markers)                    # Add the layer of markers to GMap

embed_minimal_html('MarkerMap1.html', views=[markermap])
print("*** Check your 'Metadata Part 5' folder to find the new HTML file. ***")

ModuleNotFoundError: No module named 'gmaps'